pkg/runtime/processor/binding.go

```diff
@@ -0,0 +1,135 @@
+}
+
+func (b *binding) init(ctx context.Context, comp compapi.Component) error {
+	switch {
+	case b.registry.HasInputBinding(comp.Spec.Type, comp.Spec.Version):
+		if err := b.initInputBinding(ctx, comp); err != nil {
+			return err
+		}
+
+	case b.registry.HasOutputBinding(comp.Spec.Type, comp.Spec.Version):
+		if err := b.initOutputBinding(ctx, comp); err != nil {
+			return err
+		}
+
+	default:
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return fmt.Errorf("couldn't find binding %s/%s", comp.Spec.Type, comp.Spec.Version)
+	}
+
+	return nil
+}
+
+func (b *binding) initInputBinding(ctx context.Context, comp compapi.Component) error {
+	if !b.isBindingOfDirection(bindingTypeInput, comp.Spec.Metadata) {
+		return nil
+	}
+
+	fName := comp.LogName()
+	binding, err := b.registry.CreateInputBinding(comp.Spec.Type, comp.Spec.Version, fName)
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.CreateComponentFailure, fName, err)
+	}
+	err = binding.Init(ctx, bindings.Metadata{Base: b.meta.ToBaseMetadata(comp)})
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "init", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.InitComponentFailure, fName, err)
+	}
+
+	log.Infof("successful init for input binding %s (%s/%s)", comp.ObjectMeta.Name, comp.Spec.Type, comp.Spec.Version)
+	b.compStore.AddInputBindingRoute(comp.Name, comp.Name)
+	for _, item := range comp.Spec.Metadata {
+		if item.Name == "route" {
+			b.compStore.AddInputBindingRoute(comp.ObjectMeta.Name, item.Value.String())
+			break
+		}
+	}
+	b.compStore.AddInputBinding(comp.Name, binding)
+	diag.DefaultMonitoring.ComponentInitialized(comp.Spec.Type)
+	return nil
+}
+
+func (b *binding) initOutputBinding(ctx context.Context, comp compapi.Component) error {
+	if !b.isBindingOfDirection(bindingTypeOutput, comp.Spec.Metadata) {
+		return nil
+	}
+
+	fName := comp.LogName()
+	binding, err := b.registry.CreateOutputBinding(comp.Spec.Type, comp.Spec.Version, fName)
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.CreateComponentFailure, fName, err)
+	}
+
+	if binding != nil {
+		err := binding.Init(context.TODO(), bindings.Metadata{Base: b.meta.ToBaseMetadata(comp)})
+		if err != nil {
+			diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "init", comp.ObjectMeta.Name)
+			return rterrors.NewInit(rterrors.InitComponentFailure, fName, err)
+		}
+		log.Infof("successful init for output binding %s (%s/%s)", comp.ObjectMeta.Name, comp.Spec.Type, comp.Spec.Version)
+		b.compStore.AddOutputBinding(comp.ObjectMeta.Name, binding)
+		diag.DefaultMonitoring.ComponentInitialized(comp.Spec.Type)
+	}
+	return nil
+}
+
+func (b *binding) isBindingOfDirection(direction string, metadata []common.NameValuePair) bool {
```

Comment:
This method works by looking at the component's metadata only. Shouldn't we also check if the component struct actually implements the InputBinding or OutputBinding interfaces (in addition to the metadata value)?

Reply:
I didn't write this func myself and am apprehensive about changing it's behavior here. I copied it as is. I think this should be addressed in a separate PR.

Reply:
Thanks for that. We should limit behavior change in refactor PRs.